### PR TITLE
Enable asynchronous abort exceptions during boot

### PR DESCRIPTION
**Pull request by ljerry**
On r0p2 revision and above revision, despite what is stated in ARM DDI0500D TRM
Cortex A53 r0p2 SCR_EL3 description in Table 4-77, asynchronous exceptions
generated by the code running in EL3 are not taken in EL3 when SCR_EL3[EA] is
zero.
DDI0487A.b Architecture Reference Manual does not reflect the actual
behaviour of r0p2.

DDI0487A.e Architecture Reference Manual D7.2.80 SCR_EL3 description depicts
the observed behaviour.

Fixes arm-software/tf-issues#368

Signed-off-by: Gerald Lejeune <gerald.lejeune@st.com>

### CODE REVIEW

**Comment by danh-arm** [comment]
Can one of the admins verify this patch?

**Comment by danh-arm** [comment]
Can one of the admins verify this patch?

**Comment by ljerry** [comment]
I have already "git push -f" that update the branch but not the PR.

**Comment by danh-arm** [comment]
If you mean you want to update the PR description, then you have to manually edit this. The main thing is getting the correct commit(s) visible in the PR. Is this done and ready for review?

**Comment by ljerry** [comment]
Hi Dan,

Above mentioned commit d987ad7 is available for review.

With best regards
Jerry

**Comment by danh-arm** [comment]
The commit message mentions "On r0p2 revision and above revision" but doesn't say what CPU this is talking about. Also, this is incidental information so can be removed or at least put further down the commit message.

File: bl1/aarch64/bl1_arch_setup.c [hunk 1]
```diff
@@ -38,7 +38,7 @@
 void bl1_arch_setup(void)
 {
 	/* Set the next EL to be AArch64 */
-	write_scr_el3(SCR_RES1_BITS | SCR_RW_BIT);
+	write_scr_el3(read_scr_el3() | SCR_RW_BIT);
```
Comment (danh-arm) [inline bl1/aarch64/bl1_arch_setup.c]: So you've moved initialization of the `SCR_RES1_BITS` into `el3_arch_init_common` for both BL1 and BL31, and changed this to use an SBOP (Should-Be-One-or-Preserved) policy for rewriting `SCR_EL3`. This looks correct but could do with some explanation in the commit message!

File: bl31/aarch64/crash_reporting.S [hunk 2]
```diff
@@ -338,6 +339,7 @@ func do_crash_reporting
 	mrs	x8, cntkctl_el1
 	mrs	x9, fpexc32_el2
 	mrs	x10, sp_el0
+	mrs	x11, isr_el1
```
Comment (danh-arm) [inline bl31/aarch64/crash_reporting.S]: minor: Strictly speaking this is a separate logical change and should be in a different commit. It at least needs a mention in the commit message

File: common/context_mgmt.c [hunk 3]
```diff
@@ -100,7 +100,7 @@ static void cm_init_context_common(cpu_context_t *ctx, const entry_point_info_t
 	 */
 	scr_el3 = read_scr();
 	scr_el3 &= ~(SCR_NS_BIT | SCR_RW_BIT | SCR_FIQ_BIT | SCR_IRQ_BIT |
-			SCR_ST_BIT | SCR_HCE_BIT);
+		     SCR_ST_BIT | SCR_HCE_BIT);
```
Comment (danh-arm) [inline common/context_mgmt.c]: minor: gratuitous whitespace change

File: common/context_mgmt.c [hunk 4]
```diff
@@ -111,6 +111,11 @@ static void cm_init_context_common(cpu_context_t *ctx, const entry_point_info_t
 	if (EP_GET_ST(ep->h.attr))
 		scr_el3 |= SCR_ST_BIT;
 
+#ifndef HANDLE_EA_EL3_FIRST
```
Comment (danh-arm) [inline common/context_mgmt.c]: This new build flag needs documenting in the user guide. Also, we typically allow these flags to be enabled via the build command line by adding them to the makefile and propagating them to the source via the `add_define` make function.

File: include/common/el3_common_macros.S [hunk 5]
```diff
@@ -70,10 +70,12 @@
 	isb
 
 	/* ---------------------------------------------------------------------
-	 * Enable the SError interrupt now that the exception vectors have been
+	 * Enable the Abort interrupt now that the exception vectors have been
```
Comment (danh-arm) [inline include/common/el3_common_macros.S]: This needs more clarification. We're still enabling the SError interrupt but we're also enabling SError handling at EL3.

File: include/lib/aarch64/arch.h [hunk 6]
```diff
@@ -186,6 +186,11 @@
 #define HCR_IMO_BIT		(1 << 4)
 #define HCR_FMO_BIT		(1 << 3)
 
+/* ISR definitions */
+#define ISR_A_SHIFT		8
+#define ISR_I_SHIFT		7
+#define ISR_F_SHIFT		6
+
```
Comment (danh-arm) [inline include/lib/aarch64/arch.h]: This looks like an off-topic change.

File: include/lib/aarch64/arch_helpers.h [hunk 7]
```diff
@@ -168,15 +168,6 @@ void disable_mmu_icache_el3(void);
...
 
-#define enable_irq()			write_daifclr(DAIF_IRQ_BIT)
-#define enable_fiq()			write_daifclr(DAIF_FIQ_BIT)
-#define enable_serror()			write_daifclr(DAIF_ABT_BIT)
-#define enable_debug_exceptions()	write_daifclr(DAIF_DBG_BIT)
-#define disable_irq()			write_daifset(DAIF_IRQ_BIT)
-#define disable_fiq()			write_daifset(DAIF_FIQ_BIT)
-#define disable_serror()		write_daifset(DAIF_ABT_BIT)
-#define disable_debug_exceptions()	write_daifset(DAIF_DBG_BIT)
-
```
Comment (danh-arm) [inline include/lib/aarch64/arch_helpers.h]: This looks like an off-topic change

Comment (ljerry) [inline include/lib/aarch64/arch_helpers.h]: It was conflicting with previous patch version but yes it's only code clean up regarding asynchronous exception as well. I could remove it if you still don't want it, don't you?

File: include/lib/aarch64/arch.h (hunk 6 above)
Comment (ljerry) [inline include/lib/aarch64/arch.h]: It is a mnemonic for whom is debugging such an abort with the help of ISR_EL1. I could remove it if you still don't want it, don't you?

File: common/context_mgmt.c (hunk 3 above)
Comment (ljerry) [inline common/context_mgmt.c]: Agreed

File: bl31/aarch64/crash_reporting.S (hunk 2 above)
Comment (ljerry) [inline bl31/aarch64/crash_reporting.S]: Agreed

File: bl1/aarch64/bl1_arch_setup.c (hunk 1 above)
Comment (ljerry) [inline bl1/aarch64/bl1_arch_setup.c]: Agreed

File: include/common/el3_common_macros.S (hunk 5 above)
Comment (ljerry) [inline include/common/el3_common_macros.S]: Agreed

File: include/lib/aarch64/arch_helpers.h (hunk 7 above)
Comment (danh-arm) [inline include/lib/aarch64/arch_helpers.h]: I agree that enable_serror()/disable_serror() aren't good names, and that the macros don't have much value generally. Can you put this change in a separate patch in the same Pull Request?

File: include/lib/aarch64/arch.h (hunk 6 above)
Comment (danh-arm) [inline include/lib/aarch64/arch.h]: Either put this in a separate patch (in the same PR) or remove.

Comment (ljerry) [inline include/lib/aarch64/arch.h]: ok

File: include/lib/aarch64/arch_helpers.h (hunk 7 above)
Comment (ljerry) [inline include/lib/aarch64/arch_helpers.h]: ok

**Comment by ljerry** [comment]
Hi Dan,

Above mentioned commit c82b089 8b36cd1 and e1704e2 are available for review.

With best regards
Jerry

File: include/common/el3_common_macros.S [hunk 8]
```diff
@@ -70,8 +70,15 @@
 	isb
 
 	/* ---------------------------------------------------------------------
-	 * Enable the SError interrupt now that the exception vectors have been
-	 * setup.
+	 * Early set RES1 bits in SCR_EL3. Set EA bit as well to catch both
+	 * synchronous and asynchronous exceptions in EL3.
```
Comment (danh-arm) [inline include/common/el3_common_macros.S]: Sorry to be pedantic, but I think it's really important we get the language here correct. This affects aborts, not all exception types. Also, you use the term "SError" in some places, and "asynchronous external abort" in others without saying they're the same thing. Suggest saying here "Set EA bit as well to catch External Aborts and SError Interrupts in EL3". In the commit message you could say something like "SError Interrupts are also known as asynchronous external aborts".

**Comment by danh-arm** [comment]
minor: The "Remove DAIF bits handling macros" commit message could perhaps have 1 sentence saying these macros are redundant and in the case of `enable/disable_serror()`, slightly misleading.

**Comment by danh-arm** [comment]
Otherwise, this PR looks good to me.

**Comment by danh-arm** [comment]
One more comment: I've just merged #560, which improves the firmware design document. This made me realise that we have some description there of SCR_EL3 setup, which probably needs modifying as part of your PR:
https://github.com/ARM-software/arm-trusted-firmware/blob/integration/docs/firmware-design.md#architectural-initialization

**Comment by ljerry** [comment]
Hi Dan,

New commits adb4fcf 4ca5753 and 6b836cf are available for review.

With best regards
Jerry

**Comment by danh-arm** [comment]
jenkins: test this please
